### PR TITLE
616 print image process list in nice formatted way

### LINKED ISSUE #616: Print image process list in nice formatted way
_Print image process list in nice formatted way_

**Describe the request**
Currently, image process list don't show any header and prints in not nice format list of processes (pgrep ouptut)

**Expected behavior**
Process list should show header and each line should be formatted to the expected header like:
```
~/pharo-launcher$ ./pharo-launcher.sh image processList
#  PID  Image name VM path                                                         Image path                                     Date/Time started       
-- ---- ---------- --------------------------------------------------------------- ---------------------------------------------- ------------------------
1  4291 s1         /home/dbajger/Pharo/vms/100-x64/lib/pharo                       /home/dbajger/Pharo/images/s1/s1.image         Thu May 18 13:55:19 2023
2  4338 latest     /home/dbajger/Pharo/vms/90-x64/lib/pharo/5.0-202002121043/pharo /home/dbajger/Pharo/images/latest/latest.image Thu May 18 13:55:25 2023
```

**Expected development cost**
1) Introduce new PhLImageProcess class that would include information about:
- PID, image name, VM path, image path, start date and time of process

2) Create image process object list from OS process command output
3) list the process list using ConsoleListFormatter

### PR DESCRIPTION
**Pull request by Bajger**
fixes #616

### CODE REVIEW

**Comment by Bajger** [comment]
Will add unit tests in additional commit

File: src/PharoLauncher-CLI/PhLImageProcessListCliCommand.class.st [hunk 1]
```diff
@@ -20,29 +26,46 @@ PhLImageProcessListCliCommand class >> launcherCmdDescription [
...
+PhLImageProcessListCliCommand >> imageProcesListFrom: shellOutput [
 	
-	"TODO rewrite this to more meaningful output with columns like PID, datetime started, vm name, image name"
-	"TODO: Win Powershell version should be something like this: Get-Process | Where-Object { $_.Path -like '*Pharo*' -and $_.Path -like '*.image*' -and $_.Id -ne 2232 } | Select-Object Id, Path, StartTime, CommandLine"
-	self checkOS.
-	self outStream 
-		nextPutAll: self executeOSShellCommand.
-	^ self context exitSuccess 
+	^ (shellOutput substrings: OSPlatform current lineEnding)  collect: [:line |
+			PhLImageProcess newFrom: (line splitOn: '-|-')
```
Comment (demarey) [inline src/PharoLauncher-CLI/PhLImageProcessListCliCommand.class.st]: '-|-' should be factored into a method and all occurences replaced with a call to it

File: src/PharoLauncher-CLI-Tests/ConsoleListFormatterTest.class.st [hunk 2]
```diff
@@ -294,3 +299,16 @@ ConsoleListFormatterTest >> testPrintRowValuesBy [
...
+{ #category : #tests }
+ConsoleListFormatterTest >> testTableColumnWidthLimit [
+
+	| size short longer longest|
+	short := 'short'.
+	longer := 'more long text'.
+	longest := 'some long string exceeding column width limit'.
+	size := formatter getMaxPrintStringWidthFor: {longer. longest. short.} printBlock: [ :aStr | aStr asString ].
+	
+	"three is the longest word and has a size of 5"
```
Comment (demarey) [inline src/PharoLauncher-CLI-Tests/ConsoleListFormatterTest.class.st]: I don't understand this comment

File: src/PharoLauncher-CLI-Tests/PhLImageProcessListCliCommandTest.class.st [hunk 3]
```diff
@@ -18,7 +18,28 @@ PhLImageProcessListCliCommandTest >> testExecute [
...
+		aStream 
+			nextPutAll: '1111-|-path/to/vm-|-path/to/1st.image-|-May 1 2023'; 
+			nextPutAll: OSPlatform current lineEnding; 
+			nextPutAll: '2222-|-path/to/2nd-vm-|-path/to/2nd.image-|-Jun 2 2023';
+			nextPutAll: OSPlatform current lineEnding; 
+			nextPutAll: '3333-|-path/to/3rd-vm-|-path/to/3rd.image-|-Jul 3 2023';
+			nextPutAll: OSPlatform current lineEnding
+		].
+	resultArray := processListCmd imageProcesListFrom: aString.
+	self assert: resultArray size equals: 3.
```
Comment (demarey) [inline src/PharoLauncher-CLI-Tests/PhLImageProcessListCliCommandTest.class.st]: What are we testing here?

File: src/PharoLauncher-CLI-Tests/StringTest.extension.st [hunk 4]
```diff
@@ -0,0 +1,12 @@
+Extension { #name : #StringTest }
+
+{ #category : #'*PharoLauncher-CLI-Tests' }
+StringTest >> testShortenTo [
+
+	self assert: ('ABCDEF' shortenTo: 0) equals: ' ... '.
```
Comment (demarey) [inline src/PharoLauncher-CLI-Tests/StringTest.extension.st]: ' ... ' should be factored into a method also

File: src/PharoLauncher-Core/PhLImageProcess.class.st [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+		'pid',
+		'vmPath',
+		'imagePath',
+		'startDateTime'
+	],
+	#category : #'PharoLauncher-Core-Model'
+}
+
+{ #category : #private }
+PhLImageProcess class >> listPrintAttributeBlocks [
```
Comment (demarey) [inline src/PharoLauncher-Core/PhLImageProcess.class.st:17]: Is this used?

File: src/PharoLauncher-CLI/PhLImageProcessListCliCommand.class.st (hunk 1 above)
Comment (Bajger) [reply src/PharoLauncher-CLI/PhLImageProcessListCliCommand.class.st]: Done!

File: src/PharoLauncher-CLI-Tests/ConsoleListFormatterTest.class.st (hunk 2 above)
Comment (Bajger) [reply src/PharoLauncher-CLI-Tests/ConsoleListFormatterTest.class.st]: Added comment to explain in better way.

File: src/PharoLauncher-CLI-Tests/PhLImageProcessListCliCommandTest.class.st (hunk 3 above)
Comment (Bajger) [reply src/PharoLauncher-CLI-Tests/PhLImageProcessListCliCommandTest.class.st]: Added improved comment in unit test.

File: src/PharoLauncher-Core/PhLImageProcess.class.st (hunk 5 above)
Comment (Bajger) [reply src/PharoLauncher-Core/PhLImageProcess.class.st:17]: Yes, it is used by PhLCliCommand>>listPrintAttributeBlocks that is used in console list formatter:
Line:`self modelClass listPrintAttributeBlocks`,  where modelClass is implemented by PhLImageProcessListCommand and  returns PhLImageProcess.

File: src/PharoLauncher-CLI-Tests/StringTest.extension.st (hunk 4 above)
Comment (Bajger) [reply src/PharoLauncher-CLI-Tests/StringTest.extension.st]: Moved to separate method.

**Comment by Bajger** [comment]
@demarey Hi Christophe! Finally had chance to provide needed changes. Please take a look!

**Comment by Bajger** [comment]
Found the issue in parsing line with process information. I didn't test it with pharo processes that run with --headless parameter (or other vm or image params). Therefore parsing is shifted.

**Comment by Bajger** [comment]
I need to resolve OS args passing (on MacOS), here is the situation, when no Pharo process (except launcher) is running: 
```
dbajger@Davids-MacBook-Pro P11-launcher-dev % "/Users/dbajger/Documents/Pharo/vms/110-x64/Pharo.app/Contents/MacOS/Pharo" --headless "/Users/dbajger/Documents/Pharo/images/P11-launcher-dev-proc-list/P11-launcher-dev-proc-list.image" clap launcher image processList
PhLProcessCommandError: OS process command: '/bin/zsh -c pgrep -ai -f 'pharo.*\.image' | xargs ps -o pid,args= -p | grep -v -e 'PID' | grep -v '.bash'| grep -v 'export LD_LIBRARY_PATH' | grep -v 49997'  exited with: normal termination with status 1.
Stderr contents: "".
```
but if I start some pharo image, it works: 
```
dbajger@Davids-MacBook-Pro P11-launcher-dev % "/Users/dbajger/Documents/Pharo/vms/110-x64/Pharo.app/Contents/MacOS/Pharo" --headless "/Users/dbajger/Documents/Pharo/images/P11-launcher-dev-proc-list/P11-launcher-dev-proc-list.image" clap launcher image processList
#  PID   Image name VM path                                            Image path                                         Date/Time started   
-- ----- ---------- -------------------------------------------------- -------------------------------------------------- --------------------
1  50017 P11-OTE    /Users/dbajger/Documen ... pp/Contents/MacOS/Pharo /Users/dbajger/Documen ... s/P11-OTE/P11-OTE.image 2024-01-31 08:49:07
```
Problem is in args after `/bin/zsh -c`, if no process is running then it looks like this:
```
/bin/zsh -c pgrep -ai -f 'pharo.*\.image' | xargs ps -o pid,args= -p | grep -v -e 'PID' | grep -v '.bash'| grep -v 'export LD_LIBRARY_PATH' | grep -v 49997
usage: pgrep [-Lfilnoqvx] [-d delim] [-F pidfile] [-G gid]
             [-P ppid] [-U uid] [-g pgrp] [-t tty] [-u euid]
             pattern ...
```
but if I put it in double qoutes (on terminal) it works:
```
/bin/zsh -c "pgrep -ai -f 'pharo.*\.image' | xargs ps -o pid,args= -p | grep -v -e 'PID' | grep -v '.bash'| grep -v 'export LD_LIBRARY_PATH' | grep -v 49997"

dbajger@Davids-MacBook-Pro P11-launcher-dev %
```
**So I tried to to put each arg to be surrounded by double quoutes:** 
```
osShellArgArray

	^ Array with: self processListCmdArgs surroundedByDoubleQuotes 
```
But it resulted in different error:
```
dbajger@Davids-MacBook-Pro P11-launcher-dev % "/Users/dbajger/Documents/Pharo/vms/110-x64/Pharo.app/Contents/MacOS/Pharo" --headless "/Users/dbajger/Documents/Pharo/images/P11-launcher-dev-proc-list/P11-launcher-dev-proc-list.image" clap launcher image processList
PhLProcessCommandError: OS process command: '/bin/zsh -c "pgrep -ai -f 'pharo.*\.image' | xargs ps -o pid,args= -p | grep -v -e 'PID' | grep -v '.bash'| grep -v 'export LD_LIBRARY_PATH' | grep -v 50592"'  exited with: normal termination with status 127.
Stderr contents: "zsh:1: command not found: pgrep -ai -f 'pharo.*\.image' | xargs ps -o pid,args= -p | grep -v -e 'PID' | grep -v '.bash'| grep -v 'export LD_LIBRARY_PATH' | grep -v 50592
".
```
@demarey Do you have idea, how exactly should be args qouted to OSProcess command? I pass pgrep with all stuff after as one argument (as seen in `osShellArgArray` method).

**Comment by demarey** [comment]
Hi David
From the system point of view, the command you are running is `/bin/zsh` and it has 2 arguments: `-c` and a string with all the other parts of the command. Inside the string, you have another command that will be run by `zsh`. It also has to be a valid command. The biggest issue is when you have spaces in arguments. To avoid problems, you can add single or double quotes around all arguments. To summarize, what should be done is:
- run zsh command with a string protected with single (or double) quotes
- protect arguments in the  command string that will be executed bu the shell with double quotes (could be single but not the same used for the zsh command)

**Comment by Bajger** [comment]
> run zsh command with a string protected with single (or double) quotes
> protect arguments in the command string that will be executed bu the shell with double quotes (could be single but not the same used for the zsh command)

This is what I've thought of. I've surrounded pgrep and all its args as one argument surrounded by double qoutes. So why I have and error like this? 
```
launcher image processList
PhLProcessCommandError: OS process command: '/bin/zsh -c "pgrep -ai -f 'pharo.*\.image' | xargs ps -o pid,args= -p | grep -v -e 'PID' | grep -v '.bash'| grep -v 'export LD_LIBRARY_PATH' | grep -v 50592"'  exited with: normal termination with status 127.
Stderr contents: "zsh:1: command not found: pgrep -ai -f 'pharo.*\.image' | xargs ps -o pid,args= -p | grep -v -e 'PID' | grep -v '.bash'| grep -v 'export LD_LIBRARY_PATH' | grep -v 50592
".
```
If I put exactly the same string in the terminal: `/bin/zsh -c "pgrep -ai -f 'pharo.*\.image' | xargs ps -o pid,args= -p | grep -v -e 'PID' | grep -v '.bash'| grep -v 'export LD_LIBRARY_PATH' | grep -v 50592"`
It works. 

What looks weird that `command not found`error output prints command with ending quote, but not with starting quote. Maybe some newline character could be there?

**Comment by demarey** [comment]
Maybe it has something to do with parameter expansion. In PL, the launch command does exactly what I explained below. Here is an example;
```bash
bash -c 'cd "/Users/me/Documents/Pharo/images/Pharo 11.0" && "/Users/me/Documents/Pharo/vms/110-x64/Pharo.app/Contents/MacOS/Pharo" "/Users/me/Documents/Pharo/images/Pharo 11.0/Pharo 11.0.image"'
```

**Comment by Bajger** [comment]
@demarey Please take a look. Process list identification should be robust enough now, I hope I found all cornercases for pgrep. Should be working both on Linux and MacOS (tested).
- I will squash commit and merge to feature/cmd-line
- then cherrypick to dev branch and move fiunctionality to PhLProcessListCliCommand (via new PR)?  

Note: There was a funny problem which I could not identify, since pgrep returns exitCode 1 in case of no process match. This resulted in exception in OS Shell command execution in Pharo - since exitSuccess is tested there. Lessons learned (read man pages) :-)

**Comment by demarey** [comment]
I did a local merge with dev branch and running tests.
I see a problem with date convertion:
```
OS process command: '/bin/zsh -c ps -o lstart= -p "1974" | xargs -I{} date -jf "%a %b %e %T %Y" "{}" '+%Y-%m-%d %H:%M:%S''  exited with: normal termination with status 1.
Stderr contents: "Failed conversion of ``Lun 29 jan 12:34:40 2024'' using format ``%a %b %e %T %Y''
date: illegal time format
usage: date [-jnRu] [-I[date|hours|minutes|seconds]] [-f input_fmt]
            [-r filename|seconds] [-v[+|-]val[y|m|w|d|H|M|S]]
            [[[[mm]dd]HH]MM[[cc]yy][.SS] | new_date] [+output_fmt]
".
```

**Comment by Bajger** [comment]
> `Lun 29 jan 12:34:40 2024`  
This is date in French localization? I will test with it.

**Comment by demarey** [comment]
> > `Lun 29 jan 12:34:40 2024`
> > This is date in French localization? I will test with it.

Yes. And maybe there could be other issues with otherlocalizations. Maybe there is a way to use an ISO format ?

**Comment by Bajger** [comment]
@demarey: Please try now.

**Comment by demarey** [comment]
> @demarey: Please try now.

Not yet

```
OS process command: '/bin/zsh -c LANG=C ps -o lstart= -p "1974" | xargs -I{} date -jf "%a %b %e %T %Y" "{}" '+%Y-%m-%d %H:%M:%S''  exited with: normal termination with status 1.
Stderr contents: "Failed conversion of ``Mon Jan 29 12:34:40 2024'' using format ``%a %b %e %T %Y''
date: illegal time format
usage: date [-jnRu] [-I[date|hours|minutes|seconds]] [-f input_fmt]
            [-r filename|seconds] [-v[+|-]val[y|m|w|d|H|M|S]]
            [[[[mm]dd]HH]MM[[cc]yy][.SS] | new_date] [+output_fmt]
".
```

**Comment by Bajger** [comment]
This is strange, if I evaluate on my MacOS:
```
dbajger@Davids-MacBook-Pro P11-launcher-dev % LANG=C ps -o lstart= -p 50442 | xargs -I{} date -jf "%a %b %e %T %Y" "{}" '+%Y-%m-%d %H:%M:%S' 
2024-01-31 09:07:55
```
So on my computer works. I don't know. Date formatting to ISO is really clumsy.

**Comment by demarey** [comment]
after investigation the command that works on my mac is:
` date -j -u -f "%a %d %b  %H:%M:%S %Y" '+%Y-%m-%d %H:%M:%S' "Lun 29 jan 12:26:05 2024"`

**Comment by Bajger** [comment]
@demarey Tested on my Mac with adopted time attributes (https://github.com/pharo-project/pharo-launcher/pull/617/commits/571fa76afdd7c6bdf14c38a6b69494c1505be832) . Please let me know!

**Comment by demarey** [comment]
@Bajger Did you see my PR on this topic on your repository?

**Comment by Bajger** [comment]
> @Bajger Did you see my PR on this topic on your repository?

Hi @demarey ! Sorry I didn't recognize that. To resolve conflicts, I need to checkout my branch or https://github.com/pharo-project/pharo-launcher/tree/enh/PR616 ?


<blockquote><div> GitHub</div><div><strong><a href="https://github.com/pharo-project/pharo-launcher">GitHub - pharo-project/pharo-launcher at enh/PR616</a></strong></div><div>Lets you manage your pharo images and download new ones - GitHub - pharo-project/pharo-launcher at enh/PR616</div></blockquote>

**Comment by demarey** [comment]
> Hi @demarey ! Sorry I didn't recognize that. To resolve conflicts, I need to checkout my branch or https://github.com/pharo-project/pharo-launcher/tree/enh/PR616 ?

I just diid the merge. The PR is up to date 


<blockquote><div> GitHub</div><div><strong><a href="https://github.com/pharo-project/pharo-launcher">GitHub - pharo-project/pharo-launcher at enh/PR616</a></strong></div><div>Lets you manage your pharo images and download new ones - GitHub - pharo-project/pharo-launcher at enh/PR616</div></blockquote>

**Pull request merged**